packages/ra-core/src/reducer/admin/resource/data.js

```diff
@@ -60,40 +57,43 @@ const addRecords = addRecordsFactory(getFetchedAt);
 const initialState = {};
 Object.defineProperty(initialState, 'fetchedAt', { value: {} }); // non enumerable by default
 
-export default (previousState = initialState, { type, payload, meta }) => {
-    if (type === CRUD_UPDATE_OPTIMISTIC) {
-        const updatedRecord = { ...previousState[payload.id], ...payload.data };
-        return addRecords([updatedRecord], previousState);
-    }
-    if (type === CRUD_UPDATE_MANY_OPTIMISTIC) {
-        const updatedRecords = payload.ids
-            .reduce((records, id) => records.concat(previousState[id]), [])
-            .map(record => ({ ...record, ...payload.data }));
-        return addRecords(updatedRecords, previousState);
-    }
-    if (type === CRUD_DELETE_OPTIMISTIC) {
-        const { [payload.id]: removed, ...newState } = previousState;
+export default (previousState = initialState, { payload, meta }) => {
+    if (meta && meta.optimistic) {
+        if (meta.fetch === UPDATE) {
+            const updatedRecord = { ...previousState[payload.id], ...payload.data };
+            return addRecords([updatedRecord], previousState);
+        }
+        if (meta.fetch === UPDATE_MANY) {
+            const updatedRecords = payload.ids
+                .reduce((records, id) => records.concat(previousState[id]), [])
+                .map(record => ({ ...record, ...payload.data }));
+            return addRecords(updatedRecords, previousState);
+        }
+        if (meta.fetch === DELETE) {
+            const { [payload.id]: removed, ...newState } = previousState;
 
-        Object.defineProperty(newState, 'fetchedAt', {
-            value: previousState.fetchedAt,
-        });
+            Object.defineProperty(newState, 'fetchedAt', {
+                value: previousState.fetchedAt,
+            });
 
-        return newState;
-    }
-    if (type === CRUD_DELETE_MANY_OPTIMISTIC) {
-        const newState = Object.entries(previousState)
-            .filter(([key]) => !payload.ids.includes(key))
-            .reduce((obj, [key, val]) => ({ ...obj, [key]: val }), {});
+            return newState;
+        }
+        if (meta.fetch === DELETE_MANY) {
+            const newState = Object.entries(previousState)
+                .filter(([key]) => !payload.ids.includes(key))
```

Comment:
You can put the filter logic directly in the reduce to avoid traversing the array twice

Reply:
less readable and we won't deal with many items anyway.

Reply:
👍 with @djhi here